verifiable/onchain/parser.go

```diff
@@ -0,0 +1,473 @@
+package onchain
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	core "github.com/iden3/go-iden3-core/v2"
+	"github.com/iden3/go-iden3-core/v2/w3c"
+	"github.com/iden3/go-merkletree-sql/v2"
+	"github.com/iden3/go-schema-processor/v2/merklize"
+	"github.com/iden3/go-schema-processor/v2/verifiable"
+	"github.com/piprate/json-gold/ld"
+)
+
+type Parser struct {
+	marklizeInstance merklize.Options
+
+	regenerateID bool
+}
+
+type Option func(*Parser)
+
+// WithRegenerateID sets the regenerateID flag to the given value.
+func WithRegenerateID(regenerateID bool) Option {
+	return func(p *Parser) {
+		p.regenerateID = regenerateID
+	}
+}
+
+// WithMerklizer sets the merklizer instance.
+func WithMerklizer(m merklize.Options) Option {
+	return func(p *Parser) {
+		p.marklizeInstance = m
+	}
+}
+
+func NewParser(opts ...Option) *Parser {
+	p := &Parser{
+		regenerateID: false,
+	}
+
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	return p
+}
+
+// ConvertVerifiableCredential converts an onchain verifiable credential to a W3C verifiable credential.
+// The w3c credential id will be regenerated.
+func ConvertVerifiableCredential(credential *Credential) (*verifiable.W3CCredential, error) {
+	return (&Parser{regenerateID: true}).ConvertVerifiableCredential(credential)
+}
+
+// ConvertVerifiableCredential converts an onchain verifiable credential to a W3C verifiable credential.
+func (p *Parser) ConvertVerifiableCredential(onchainCredential *Credential) (*verifiable.W3CCredential, error) {
+	timestampExp, err := strconv.ParseInt(onchainCredential.Expiration, 10, 64)
+	if err != nil {
+		return nil,
+			fmt.Errorf("failed to parse Expiration '%s': %v", onchainCredential.Expiration, err)
+	}
+	timestampIssuance, err := strconv.ParseInt(onchainCredential.IssuanceDate, 10, 64)
+	if err != nil {
+		return nil,
+			fmt.Errorf("failed to parse IssuanceDate '%s': %v", onchainCredential.IssuanceDate, err)
+	}
+	expirationTime := time.Unix(timestampExp, 0).UTC()
```

Comment:
how to handle usecases when timestamp is not seconds, but nanoseconds (e.g. Quorum blockchain)

The issue is that Ethereum returns the timestamp in seconds, but Quorum returns a timestamp in nanoseconds.

Reply:
I agree need to discuss how we can process it.
But we can create an option for the parser/converter. WithNanosecondsTimestamp,